sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatRecordReaderFn.java

```diff
@@ -0,0 +1,165 @@
+      p.add(partition);
+      final List<Long> fileSizeForPartitions = StatsUtils.getFileSizeForPartitions(hiveConf, p);
+      return fileSizeForPartitions.get(0);
+    } finally {
+      // IMetaStoreClient is not AutoCloseable, closing it manually
+      if (client != null) {
+        client.close();
+      }
+    }
+  }
+
+  /** Reads data for a specific partition. */
+  @ProcessElement
+  @SuppressWarnings("unused")
+  public void processElement(
+      ProcessContext processContext, RestrictionTracker<OffsetRange, Long> tracker)
+      throws Exception {
+    final Read readRequest = processContext.element();
+    final List<String> values = readRequest.getPartitionToRead().getValues();
+    final String watermarkPartitionColumn = readRequest.getWatermarkPartitionColumn();
+    final int indexOfPartitionColumnWithWaterMark =
+        watermarkPartitionColumn.indexOf(watermarkPartitionColumn);
+    final String partitionWaterMark = values.get(indexOfPartitionColumnWithWaterMark);
+
+    final Instant partitionWatermarkTimeStamp =
+        readRequest.getWatermarkTimestampConverter().apply(partitionWaterMark);
+
+    int desiredSplitCount = 1;
+    long estimatedSizeBytes = getFileSizeForPartitions(processContext.element());
+    if (estimatedSizeBytes > 0) {
+      desiredSplitCount = (int) Math.ceil((double) estimatedSizeBytes / Integer.MAX_VALUE);
+    }
+    ReaderContext readerContext = getReaderContext(processContext.element(), desiredSplitCount);
+
+    for (long i = tracker.currentRestriction().getFrom(); tracker.tryClaim(i); i++) {
+      HCatReader reader = DataTransferFactory.getHCatReader(readerContext, (int) i);
+      Iterator<HCatRecord> hcatIterator = reader.read();
+      while (hcatIterator.hasNext()) {
+        final HCatRecord record = hcatIterator.next();
+        processContext.output(record);
+      }
+    }
+    LOG.info("Watermark update to {}", partitionWatermarkTimeStamp);
+    // Once we have read completely from partition, we can advance the watermark associated with the
+    // partition
+    processContext.updateWatermark(partitionWatermarkTimeStamp);
+  }
+
+  @GetInitialRestriction
+  @SuppressWarnings("unused")
+  public OffsetRange getInitialRestriction(Read readRequest) throws Exception {
+    int desiredSplitCount = 1;
+    long estimatedSizeBytes = getFileSizeForPartitions(readRequest);
+    if (estimatedSizeBytes > 0) {
+      desiredSplitCount = (int) Math.ceil((double) estimatedSizeBytes / Integer.MAX_VALUE);
+    }
+    ReaderContext readerContext = getReaderContext(readRequest, desiredSplitCount);
+    // process the splits returned by native API
+    // this could be different from 'desiredSplitCount' calculated above
+    LOG.info(
+        "Splitting: estimated size {}, desired split count {}, actual split count",
+        estimatedSizeBytes,
+        desiredSplitCount);
+    return new OffsetRange(0, readerContext.numSplits());
+  }
+
+  @NewTracker
```

Comment:
This is unnecessary to declare, the default implementation will do the same thing.